README.md
# ESPBASIC
BASIC for the ESP32. FabGL is used as the graphics engine.
<br>
Currently still in beta. Version as of 2020-10-14 (ISO 8601) is 0.0.0.29.
<br>
Read changelog.txt for more details.


changelog.txt
Beta:
v0.0.0.29:
  added more hi-res video modes (must run in 16 colors and may have to change mas = 65536; to mas = 49152; to use vmode 6 and 13)
v0.0.0.28:
  made floating point calculations as accurate as I could
  added 16 color mode (default due to speed increase)
v0.0.0.27:
  fixed number variable store/recall
  added integer number type
  fixed weird omptimization-caused number wrapping bug
  now using VICNoiseGenerator instead of NoiseWaveformGenerator
v0.0.0.26:
  bug fixes
  added BASIC command: SOUND (can also be accessed using SND)
  added BASIC command: DELAY
  fixed prompt newline glitch
v0.0.0.25:
  bug fixes
v0.0.0.24:
  bug fixes
  added BASIC command: DO
  added BASIC command: LOOP
  added multi-command lines (commands are separated by ':')
v0.0.0.23:
  bug fixes
v0.0.0.22:
  fixed number interpreter bug that made calculating negative numbers impossible
  added BASIC command: COLOR (COLOR fgc [, bgc])
  changed syntax of BASIC command CLS (accepts an optional color value)
  changed syntax of BASIC command PRINT (PRINT VALUE/VALUE$ -> PRINT VALUE/VALUE$ [,/; VALUE/VALUE$ ...])
  i: I plan on adding the ability to put more that one command on a line separated by ':'
v0.0.0.21:
  added variable making ability (VARIABLE=VALUE)
  !: Known bug: Setting a number variable will crash the ESP-32 with "Guru Meditation Error: Core  1 panic'ed (LoadStoreError). Exception was unhandled."
  fixed varible storage bugs
  added memory saving routine (Though it is not fast, it works. I plan on adding some options to this when the SYS command is added.)
v0.0.0.20:
  fixed bugs
  added number interpreter
  moved from alpha to beta
Alpha:
v0.0.0.19:
  redid the string interpreter
v0.0.0.18:
  fixed variable storage bug
v0.0.0.17:
  added video multiple video modes
  added BASIC command: VMODE0
  added BASIC command: VMODE1
  added BASIC command: VMODE2
v0.0.0.16:
  fixed expression interpreter bugs
v0.0.0.15:
  fixed string interpreter bugs
v0.0.0.14:
  fixed expression interpreter bugs
  added fully functional string interpreter
  added sound functionality
  added BASIC command: SNDTEST
  split off extra commands into separate file
  added startup chime sound test:
    F1 = Square wave
    F2 = Triangle wave
    F3 = Sawtooth wave
    F4 = Sine wave
    F5 = Noise wave
v0.0.0.13:
  fixed bugs in dummy terminal
v0.0.0.12:
  improved syntax detection
v0.0.0.11:
  added half of an expression interpreter
  improved dummy terminal
  improved keyboard routines
  fixed prompt home key line beginning offset
  added BASIC command: PRINT (can also be accessed using ?)
  added BASIC command: SYSINFO (can also be accessed using SYSINF)
  renamed code from sketch_jul02a.ino to ESPBASIC.ino
v0.0.0.10:
  improved dummy terminal
  added BASIC command: CLS
  added BASIC command: MEMORY (can also be accessed using MEM)
  replaced BASIC command EXIT with DUMMY due to future conflicts
  improved BASIC prompt
v0.0.0.9:
  added BASIC functionality
  added BASIC command: EXIT
v0.0.0.8:
  working variable save and recall for strings
  changed variable save technique
  text at startup uses variable r/w routines
v0.0.0.7:
  improved variable save and recall
  added some workarounds
v0.0.0.6:
  changed some routines
v0.0.0.5:
  added some keyboard routines
  added software reset by CTRL+ALT+DEL
v0.0.0.4:
  used different RAM allocating technique
  improved dummy terminal
  added more features to the dummy terminal
v0.0.0.3:
  retracted RAM managment routines
v0.0.0.2:
  enhanced dummy terminal
  added rudimentary RAM management
v0.0.0.1:
  added dummy terminal


funCMDs.h
// ESPBASIC eXtra commands v0.0.0.16
///*
//Uncomment this section to enable the crash command
if (CMD == "CRASH") {
  sicon(3); printString("Are you sure?");
ck:
  rfKB();
  drawCursor(tcx, tcy);
  if (chr == 'y' || chr == 'Y') {
    goto oof;
  } else if (chr != 0) {
    goto whew;
  }
  goto ck;
oof:
  randomSeed(timer());
  long rip;
  do {
    rip = random(-98304, 165940);
    rfKB();
    drawCursor(tcx, tcy);
    pmem[rip] = random(0, 256 * chr);
    tbfr[random(-16, 65)][random(-4, 41)] = random(0, 256);
    GFX.setPenColor(random(0, 256), random(0, 256), random(0, 256));
    GFX.setBrushColor(random(0, 256), random(0, 256), random(0, 256));
    GFX.drawChar(random(-16384, 16384), random(-16384, 16384), random(0, 256));
    setFGColor(fgc);
    setBGColor(bgc);
    rfScrTxt();
  } while (true);
whew:
  printChar(13);
  nac = false;
}
//*/
///*
//Uncomment this section to enable the C64 theme command
if (CMD == "THEME64") {
  fgc = 43;
  bgc = 18;
  setFGColor(fgc);
  setBGColor(bgc);
  rfScrTxt();
  nac = false;
}
//*/
///*
//Uncomment this section to enable the C128 theme command
if (CMD == "THEME128") {
  fgc = 46;
  bgc = 21;
  setFGColor(fgc);
  setBGColor(bgc);
  rfScrTxt();
  nac = false;
}
//*/
///*
//Uncomment this section to enable the default theme command
if (CMD == "THEMEDF") {
  fgc = 60;
  bgc = 1;
  setFGColor(fgc);
  setBGColor(bgc);
  rfScrTxt();
  nac = false;
}
//*/
///*
//Uncomment this section to enable the b&w theme command
if (CMD == "THEMEBW") {
  fgc = 42;
  bgc = 0;
  setFGColor(fgc);
  setBGColor(bgc);
  rfScrTxt();
  nac = false;
}
//*/
///*
//Uncomment this section to enable the b&w bright theme command
if (CMD == "THEMEBWB") {
  fgc = 63;
  bgc = 0;
  setFGColor(fgc);
  setBGColor(bgc);
  rfScrTxt();
  nac = false;
}
//*/


FunCMDs.h
// ESPBASIC eXtra commands v0.0.0.29
///*
//Uncomment this section to enable the crash command
if (CMD == "CRASH") {
  sicon(3); prompt("Are you sure?: ");
  lineString = line; lineString.trim(); lineString.toUpperCase();
  if (lineString == "Y" || lineString == "YES") {
    goto oof;
  } else {
    goto whew;
  }
oof:
  randomSeed(timer());
  long rip;
  while (chr != 3) {
    rip = random(-98304, 165940);
    rfKB();
    drawCursor(tcx, tcy);
    pmem[rip] = random(0, 256 * chr);
    tbfr[random(-16, 65)][random(-4, 41)] = random(0, 256);
    GFX.setPenColor(random(0, 256), random(0, 256), random(0, 256));
    GFX.setBrushColor(random(0, 256), random(0, 256), random(0, 256));
    GFX.drawChar(random(-16384, 16384), random(-16384, 16384), random(0, 256));
    setFGColor(fgc);
    setBGColor(bgc);
    rfScrTxt();
  }
whew:
  printChar(13);
  nac = false;
}
//*/
///*
//Uncomment this section to enable the C64 theme command
if (CMD == "THEME64") {
  fgc = 43;
  bgc = 18;
  setFGColor(fgc);
  setBGColor(bgc);
  rfScrTxt();
  nac = false;
}
//*/
///*
//Uncomment this section to enable the C128 theme command
if (CMD == "THEME128") {
  fgc = 46;
  bgc = 21;
  setFGColor(fgc);
  setBGColor(bgc);
  rfScrTxt();
  nac = false;
}
//*/
///*
//Uncomment this section to enable the default theme command
if (CMD == "THEMEDF") {
  fgc = 60;
  bgc = 1;
  setFGColor(fgc);
  setBGColor(bgc);
  rfScrTxt();
  nac = false;
}
//*/
///*
//Uncomment this section to enable the b&w theme command
if (CMD == "THEMEBW") {
  fgc = 42;
  bgc = 0;
  setFGColor(fgc);
  setBGColor(bgc);
  rfScrTxt();
  nac = false;
}
//*/
///*
//Uncomment this section to enable the b&w bright theme command
if (CMD == "THEMEBWB") {
  fgc = 63;
  bgc = 0;
  setFGColor(fgc);
  setBGColor(bgc);
  rfScrTxt();
  nac = false;
}
//*/
